File: backend/src/main/java/org/luzkix/coinchange/service/UserService.java
package org.luzkix.coinchange.service;

import org.luzkix.coinchange.model.User;
import org.luzkix.coinchange.openapi.backendapi.model.*;

import java.util.List;

public interface UserService {
    UserLoginResponseDto createUser(UserRegistrationRequestDto registrationDto);

    UserLoginResponseDto loginUser(UserLoginRequestDto userLoginDto);

    User findUserById(Long userId);

    List<User> findAllActiveUsers();

    RefreshTokenResponseDto refreshToken(User user);

    UserLoginResponseDto updateUser(UserUpdateRequestDto userLoginDto, User user);

    UserLoginResponseDto suspendUser(User user);
}

File: frontend/src/services/UserService.ts
import {
  ApiUserService,
  RefreshTokenResponseDto,
  UserLoginRequestDto,
  UserLoginResponseDto,
  UserRegistrationRequestDto,
  UserUpdateRequestDto,
} from '../api-generated/backend';
import { RefreshTokenError } from '../constants/customErrors.ts';

/**
 * UserService provides api calls related to user administration.
 */
export const UserService = {
  /**
   * Registers a new user using the backend API.
   * Returns UserLoginResponseDto on success.
   */
  async register(data: UserRegistrationRequestDto): Promise<UserLoginResponseDto> {
    console.log('Registering user...');
    return ApiUserService.createUser(data);
  },
  /**
   * Logs-in a user using the backend API.
   * Returns UserLoginResponseDto on success.
   */
  async login(data: UserLoginRequestDto): Promise<UserLoginResponseDto> {
    console.log(`Logging-in the user with username/email: ${data.usernameOrEmail} ...`);
    return ApiUserService.loginUser(data);
  },
  /**
   * Getting fresh token with updated validity using the backend API.
   * Returns RefreshTokenResponseDto on success.
   */
  async refreshToken(): Promise<RefreshTokenResponseDto> {
    try {
      console.log('Refreshing user token...');
      return await ApiUserService.refreshToken();
    } catch (error) {
      const message = `Failed to refresh user token`;
      console.log(message);

      throw new RefreshTokenError(
        message,
        error instanceof Error ? error : new Error(String(error)),
      );
    }
  },
  /**
   * Updates a user using the backend API.
   * Returns UserLoginResponseDto on success.
   */
  async update(data: UserUpdateRequestDto): Promise<UserLoginResponseDto> {
    console.log('Updating user data...');
    return ApiUserService.updateUser(data);
  },
};
